# Data exploration and cleaning

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Data loading and exploration

In [2]:
orders = pd.read_csv('../00.Data/orders_cripted.csv')

In [3]:
orders.head()

Name Financial Status                    Paid at Fulfillment Status  \
0  #1244             paid  2019-04-15 09:51:49 +0200        unfulfilled   
1  #1243             paid  2019-04-11 23:35:23 +0200          fulfilled   
2  #1243              NaN                        NaN                NaN   
3  #1242         refunded  2019-04-11 23:21:35 +0200        unfulfilled   
4  #1242              NaN                        NaN                NaN   

                Fulfilled at Accepts Marketing Currency  Subtotal  Shipping  \
0                        NaN               yes      EUR      45.0       4.9   
1  2019-04-11 23:43:20 +0200               yes      EUR      40.1       4.9   
2                        NaN               NaN      NaN       NaN       NaN   
3                        NaN               yes      EUR      40.1       4.9   
4                        NaN               NaN      NaN       NaN       NaN   

   Taxes  ...  Tax 1 Value Tax 2 Name  Tax 2 Value Tax 3 Name Tax 3 Value  \
0    0.0  ...          NaN        NaN          NaN        NaN         NaN   
1    0.0  ...          NaN        NaN          NaN        NaN         NaN   
2    NaN  ...          NaN        NaN          NaN        NaN         NaN   
3    0.0  ...          NaN        NaN          NaN        NaN         NaN   
4    NaN  ...          NaN        NaN          NaN        NaN         NaN   

   Tax 4 Name Tax 4 Value  Tax 5 Name  Tax 5 Value  Receipt Number  
0         NaN         NaN         NaN          NaN             NaN  
1         NaN         NaN         NaN          NaN             NaN  
2         NaN         NaN         NaN          NaN             NaN  
3         NaN         NaN         NaN          NaN             NaN  
4         NaN         NaN         NaN          NaN             NaN  

[5 rows x 49 columns]

It seems that we have a lot of nan values and I'm not sure what the Tax N Value and Name are and seem to be empty. Let's check all Nan values compared to the total amount of data we have.

In [4]:
orders.shape

(2415, 49)

In [5]:
orders.isna().sum()

Name                              0
Financial Status               2178
Paid at                        2258
Fulfillment Status             2178
Fulfilled at                   2208
Accepts Marketing              2178
Currency                       2178
Subtotal                       2178
Shipping                       2178
Taxes                          2178
Total                          2178
Discount Code                  2378
Discount Amount                2178
Shipping Method                2188
Created at                        0
Lineitem quantity                 0
Lineitem name                     0
Lineitem price                    0
Lineitem compare at price      2392
Lineitem sku                     79
Lineitem requires shipping        0
Lineitem taxable                  0
Lineitem fulfillment status       0
Billing City                   2186
Billing Zip                    2186
Billing Country                2185
Notes                          2362
Note Attributes             

Ok, so we have plenty of NaNs. We'll drop all 'Tax' related values except the #1 as there is no data in them. Also 'Receipt Number' has no values so we will also drop it.

In [6]:
orders.drop(columns=['Tax 2 Name', 'Tax 2 Value', 'Tax 3 Name', 'Tax 3 Value', 'Tax 4 Name', 'Tax 4 Value', 'Tax 5 Name', 'Tax 5 Value', 'Receipt Number'], inplace=True)

Each row is a register for an order but there's no unique Id for each row. The 'Id' we have is mostly empty. 
We have 2 possible scenarios:

1) 'Id' is the order id and we only have 2415 - 2178 = 237 orders

2) All the rows are actually orders but there is no id.

The most probable option is 1 but to discard option 2, we will group by 'Created at' as it is a not-null variable and if two rows are created at exactly the same time, we can assume that correspond to the same order.

In [7]:
orders.groupby('Created at').count().shape

(237, 39)

We have 237 different Timestamps, meaning that we actually have 237 orders, the empty data refers to specific information of the order. Actually, the 'Name' appears to be the ID of the orders.

In [8]:
orders.groupby('Name').count().shape

(237, 39)

In [9]:
orders['Name'].value_counts().head()

#1164    34
#1171    33
#1051    32
#1157    30
#1153    30
Name: Name, dtype: int64

In [10]:
orders['Name'].value_counts().tail()

#1214    1
#1185    1
#1187    1
#1044    1
#1203    1
Name: Name, dtype: int64

So we have some orders with 34 rows of information and some with just 1. Let's check the order #1164 to better understand what type of information we might find in the additional rows.

In [11]:
orders[orders['Name']=='#1164']

Name Financial Status                    Paid at Fulfillment Status  \
266  #1164             paid  2019-01-15 22:58:00 +0100          fulfilled   
267  #1164              NaN                        NaN                NaN   
268  #1164              NaN                        NaN                NaN   
269  #1164              NaN                        NaN                NaN   
270  #1164              NaN                        NaN                NaN   
271  #1164              NaN                        NaN                NaN   
272  #1164              NaN                        NaN                NaN   
273  #1164              NaN                        NaN                NaN   
274  #1164              NaN                        NaN                NaN   
275  #1164              NaN                        NaN                NaN   
276  #1164              NaN                        NaN                NaN   
277  #1164              NaN                        NaN                NaN   
278  #1164              NaN                        NaN                NaN   
279  #1164              NaN                        NaN                NaN   
280  #1164              NaN                        NaN                NaN   
281  #1164              NaN                        NaN                NaN   
282  #1164              NaN                        NaN                NaN   
283  #1164              NaN                        NaN                NaN   
284  #1164              NaN                        NaN                NaN   
285  #1164              NaN                        NaN                NaN   
286  #1164              NaN                        NaN                NaN   
287  #1164              NaN                        NaN                NaN   
288  #1164              NaN                        NaN                NaN   
289  #1164              NaN                        NaN                NaN   
290  #1164              NaN                        NaN                NaN   
291  #1164              NaN                        NaN                NaN   
292  #1164              NaN                        NaN                NaN   
293  #1164              NaN                        NaN                NaN   
294  #1164              NaN                        NaN                NaN   
295  #1164              NaN                        NaN                NaN   
296  #1164              NaN                        NaN                NaN   
297  #1164              NaN                        NaN                NaN   
298  #1164              NaN                        NaN                NaN   
299  #1164              NaN                        NaN                NaN   

                  Fulfilled at Accepts Marketing Currency  Subtotal  Shipping  \
266  2019-01-15 23:46:53 +0100               yes      EUR     105.1       0.0   
267                        NaN               NaN      NaN       NaN       NaN   
268                        NaN               NaN      NaN       NaN       NaN   
269                        NaN               NaN      NaN       NaN       NaN   
270                        NaN               NaN      NaN       NaN       NaN   
271                        NaN               NaN      NaN       NaN       NaN   
272                        NaN               NaN      NaN       NaN       NaN   
273                        NaN               NaN      NaN       NaN       NaN   
274                        NaN               NaN      NaN       NaN       NaN   
275                        NaN               NaN      NaN       NaN       NaN   
276                        NaN               NaN      NaN       NaN       NaN   
277                        NaN               NaN      NaN       NaN       NaN   
278                        NaN               NaN      NaN       NaN       NaN   
279                        NaN               NaN      NaN       NaN       NaN   
280                        NaN               NaN      NaN       NaN       NaN   
281           

We can see that there is unique information for the first row. This would be the general order information, which includes: 

    order_specific = ['Name', 'Financial Status', 'Paid at', 'Fulfillment Status', 'Fulfilled at', 'Accepts Marketing', 'Currency', 'Subtotal', 'Shipping', 'Taxes', 'Total', 'Discount Amount', 'Shipping Method', 'Billing City', 'Billing Zip', 'Billing Country', 'Note Attributes', 'Payment Method', 'Payment Reference', 'Refunded Amount', 'Id', 'Risk Level', 'Source']

Other variables are clearly item-specific:

    item_specific = ['Lineitem quantity', 'Lineitem name', 'Lineitem price', 'Lineitem sku', 'Lineitem requires shipping', 'Lineitem taxable', 'Lineitem fulfillment status']
    
    
Some variables could be either also order-specific or item-specific information but need to be confirmed as here they contain NaN values for all the rows.

    uncertain_order = ['Discount Code','Notes', 'Cancelled at', 'Tags', 'Tax 1 Name', 'Tax 1 Value']
    uncertain_item = ['Lineitem compare at price']

We will now split the DataFrame as follows:
* order_specific information
* item_specific information

To do that we need:

1) Classify the uncertain variables

2) Reference both tables with some sort of Order Id

### Classify the uncertain variables

In [12]:
orders[['Discount Code','Notes', 'Cancelled at', 'Tags', 'Tax 1 Name', 'Tax 1 Value']].notna().sum()

Discount Code    37
Notes            53
Cancelled at     29
Tags              4
Tax 1 Name        6
Tax 1 Value       6
dtype: int64

In [13]:
orders['Discount Code'].value_counts()

PRIMERACOMPRA     12
LAERAVIMET         9
RECLIENT02         2
NEWSLETTER         2
#PRIMERACESTA      2
#NUEVOVIMET        1
#PILI              1
CESTAVIMET         1
RECLIENT06         1
#RECAPTE           1
#MARIAVIMET        1
#PRIMERACOMPRA     1
TeamPati           1
CLIENTEVIP         1
#MARIAC            1
Name: Discount Code, dtype: int64

In [14]:
orders[orders['Discount Code'].notna()]['Name'].value_counts().head()

#1233    1
#1149    1
#1110    1
#1073    1
#1126    1
Name: Name, dtype: int64

We can see that 'Discount Code' is unique per order. We'll do the same for the rest of variables

In [15]:
orders[orders['Notes'].notna()]['Name'].value_counts().head()

#1022    1
#1215    1
#1056    1
#1236    1
#1220    1
Name: Name, dtype: int64

In [16]:
orders[orders['Cancelled at'].notna()]['Name'].value_counts().head()

#1022    1
#1182    1
#1176    1
#1007    1
#1049    1
Name: Name, dtype: int64

In [17]:
orders[orders['Tags'].notna()]['Name'].value_counts().head()

#1234    1
#1232    1
#1235    1
#1233    1
Name: Name, dtype: int64

In [18]:
orders[orders['Tax 1 Name'].notna()]['Name'].value_counts().head()

#1065    1
#1075    1
#1174    1
#1176    1
#1199    1
Name: Name, dtype: int64

In [19]:
orders[orders['Tax 1 Value'].notna()]['Name'].value_counts().head()

#1065    1
#1075    1
#1174    1
#1176    1
#1199    1
Name: Name, dtype: int64

In [20]:
orders[orders['Lineitem compare at price'].notna()]['Name'].value_counts().head()

#1109    2
#1165    2
#1152    2
#1134    2
#1164    1
Name: Name, dtype: int64

All the uncertain variables are unique to each order or item as we suspected (so uncertain lists are now certain) but only fulfilled for few orders/items. We will put them in the specific DataFrames and then decide later what to do with it. The updated list of variables per dataframe is the following:

    order_specific = ['Name', 'Financial Status', 'Paid at', 'Fulfillment Status', 'Fulfilled at', 'Accepts Marketing', 'Currency', 'Subtotal', 'Shipping', 'Taxes', 'Total', 'Discount Amount', 'Shipping Method', 'Billing City', 'Billing Zip', 'Billing Country', 'Note Attributes', 'Payment Method', 'Payment Reference', 'Refunded Amount', 'Id', 'Risk Level', 'Source', 'Discount Code','Notes', 'Cancelled at', 'Tags', 'Tax 1 Name', 'Tax 1 Value']

    item_specific = ['Lineitem quantity', 'Lineitem name', 'Lineitem price', 'Lineitem sku', 'Lineitem requires shipping', 'Lineitem taxable', 'Lineitem fulfillment status', 'Lineitem compare at price']

## Splitting the dataframe

To ease the understanding of the information we will split our dataframe in 2. The first will be order-specific and the second item-specific. This will help us to better structure the database, and clean each DF separately with a better undestanding of each variable.

In [21]:
cols_order = ['Name', 'Financial Status', 'Paid at', 'Fulfillment Status', 'Fulfilled at', 'Accepts Marketing', 'Currency', 'Subtotal', 'Shipping', 'Taxes', 'Total', 'Discount Amount', 'Shipping Method', 'Billing City', 'Billing Zip', 'Billing Country', 'Note Attributes', 'Payment Method', 'Payment Reference', 'Refunded Amount', 'Id', 'Risk Level', 'Source', 'Discount Code','Notes', 'Cancelled at', 'Tags', 'Tax 1 Name', 'Tax 1 Value']
order_specific = orders[cols_order]

#note that we are including 'name' and 'id' to the item-specific dataframe so that this information is traceable to the specific order.
cols_item = ['Name', 'Id', 'Lineitem quantity', 'Lineitem name', 'Lineitem price', 'Lineitem sku', 'Lineitem requires shipping', 'Lineitem taxable', 'Lineitem fulfillment status', 'Lineitem compare at price']
item_specific = orders[cols_item]

## Cleaning order-specific data

In [22]:
order_specific.head()

Name Financial Status                    Paid at Fulfillment Status  \
0  #1244             paid  2019-04-15 09:51:49 +0200        unfulfilled   
1  #1243             paid  2019-04-11 23:35:23 +0200          fulfilled   
2  #1243              NaN                        NaN                NaN   
3  #1242         refunded  2019-04-11 23:21:35 +0200        unfulfilled   
4  #1242              NaN                        NaN                NaN   

                Fulfilled at Accepts Marketing Currency  Subtotal  Shipping  \
0                        NaN               yes      EUR      45.0       4.9   
1  2019-04-11 23:43:20 +0200               yes      EUR      40.1       4.9   
2                        NaN               NaN      NaN       NaN       NaN   
3                        NaN               yes      EUR      40.1       4.9   
4                        NaN               NaN      NaN       NaN       NaN   

   Taxes  ...  Refunded Amount            Id Risk Level Source Discount Code  \
0    0.0  ...              0.0  8.910100e+11        Low    web           NaN   
1    0.0  ...              0.0  8.869630e+11        Low    web           NaN   
2    NaN  ...              NaN           NaN        NaN    NaN           NaN   
3    0.0  ...             45.0  8.869480e+11        Low    web           NaN   
4    NaN  ...              NaN           NaN        NaN    NaN           NaN   

  Notes               Cancelled at Tags Tax 1 Name  Tax 1 Value  
0   NaN                        NaN  NaN        NaN          NaN  
1   NaN                        NaN  NaN        NaN          NaN  
2   NaN                        NaN  NaN        NaN          NaN  
3   NaN  2019-04-11 23:32:02 +0200  NaN        NaN          NaN  
4   NaN                        NaN  NaN        NaN          NaN  

[5 rows x 29 columns]

We need to keep only the first row of each order, the rest contains only NaN values.

In [23]:
order_specific_clean = pd.DataFrame()
for order in list(order_specific['Name'].unique()):
    order_specific_clean = order_specific_clean.append(order_specific[order_specific['Name']==order].head(1))

In [24]:
order_specific_clean.isna().sum()

Name                    0
Financial Status        0
Paid at                80
Fulfillment Status      0
Fulfilled at           30
Accepts Marketing       0
Currency                0
Subtotal                0
Shipping                0
Taxes                   0
Total                   0
Discount Amount         0
Shipping Method        10
Billing City            8
Billing Zip             8
Billing Country         7
Note Attributes        17
Payment Method          0
Payment Reference      37
Refunded Amount         0
Id                      0
Risk Level              0
Source                  0
Discount Code         200
Notes                 184
Cancelled at          208
Tags                  233
Tax 1 Name            231
Tax 1 Value           231
dtype: int64

We can see that:
* 'Paid at': 80 orders were not registered as paid
* 'Fullfilled at': 30 orders were not fulfilled
* 'Shipping method': 10 orders do not have shipping method
* 'Billing city' and 'Billing zip': 8 orders do not have Billing city nor Billing Zip
* 'Billing country': 7 orders do not have Billing country
* 'Note attributes': 17 orders do not have Note Attributes
* 'Payment reference': 37 orders do not have a payment reference
* 'Discount Code': 200 orders do not have a discout code
* 'Notes': 184 orders do not have notes
* 'Canelled at': 208 orders were not cancelled
* 'Tags': 233 orders do not have tags
* 'Tax Name' and 'Tax Value': 231 orders do not have tax information

We will now analyse each of the variables with NaN values to correct this issue.

### Order-specific data: Solving NaN values

#### Dropping values

We will drop the following variables as do not provide relevant information for the analysis:

'Paid at', 'Fullfilled at', 'Billing Country', 'Note Attributes', 'Payment Reference', 'Cancelled at', 'Tags', 'Tax 1 Name', 'Tax 1 Value'

In [25]:
order_specific_clean = order_specific_clean.drop(columns=['Paid at', 'Fulfilled at', 'Billing Country', 'Note Attributes', 'Payment Reference', 'Cancelled at', 'Tags', 'Tax 1 Name', 'Tax 1 Value'])



In [26]:
#We check the status of the NaN values
order_specific_clean.isna().sum()

Name                    0
Financial Status        0
Fulfillment Status      0
Accepts Marketing       0
Currency                0
Subtotal                0
Shipping                0
Taxes                   0
Total                   0
Discount Amount         0
Shipping Method        10
Billing City            8
Billing Zip             8
Payment Method          0
Refunded Amount         0
Id                      0
Risk Level              0
Source                  0
Discount Code         200
Notes                 184
dtype: int64

We don't care about the discount code but we do care whether there was a discount or not.

In [27]:
order_specific_clean['Discount Code'] = order_specific_clean['Discount Code'].fillna(value=0)

In [28]:
order_specific_clean['Discount Code'] = order_specific_clean['Discount Code'].map(lambda x: 0 if x==0 else 1)

We need to analyse 'Notes' as it will identify test records

In [29]:
order_specific_clean['Notes'] = order_specific_clean['Notes'].fillna(value='None')
order_specific_clean['Notes'] = order_specific_clean['Notes'].map(lambda x: 'test' if 'prueba' in x else ('test' if 'prova' in x else ('test' if 'Hola' in x else ('test' if 'fdfsdfsdf' in x else ('test' if 'wqweqwe' in x else ('test' if 'hohsadasdasd' in x else ('test' if 'fsdfsfsd' in x else ('test' if 'hola mundo' in x else x))))))))


We now remove all test records

In [30]:
order_specific_clean = order_specific_clean[order_specific_clean['Notes'] != 'test']

In [31]:
#We check the status of the NaN values
order_specific_clean.isna().sum()

Name                   0
Financial Status       0
Fulfillment Status     0
Accepts Marketing      0
Currency               0
Subtotal               0
Shipping               0
Taxes                  0
Total                  0
Discount Amount        0
Shipping Method       10
Billing City           8
Billing Zip            8
Payment Method         0
Refunded Amount        0
Id                     0
Risk Level             0
Source                 0
Discount Code          0
Notes                  0
dtype: int64

In [32]:
order_specific_clean['Shipping Method'].value_counts()

Envío estándar                              133
Envío gratuito                               51
Recogida en el mercado                       21
El reparto se realiza a través de Shargo      4
Name: Shipping Method, dtype: int64

When 'Shipping Method' is NaN, it means that it is delivered by one of the employees. We will fill NaNs with 'Other'

In [33]:
order_specific_clean['Shipping Method'] = order_specific_clean['Shipping Method'].fillna(value='Other')

In [34]:
#We check the status of the NaN values
order_specific_clean.isna().sum()

Name                  0
Financial Status      0
Fulfillment Status    0
Accepts Marketing     0
Currency              0
Subtotal              0
Shipping              0
Taxes                 0
Total                 0
Discount Amount       0
Shipping Method       0
Billing City          8
Billing Zip           8
Payment Method        0
Refunded Amount       0
Id                    0
Risk Level            0
Source                0
Discount Code         0
Notes                 0
dtype: int64

In [35]:
order_specific_clean[order_specific_clean['Billing City'].isna()]

Name Financial Status Fulfillment Status Accepts Marketing Currency  \
5   #1241             paid          fulfilled                no      EUR   
24  #1231             paid          fulfilled                no      EUR   
36  #1228             paid          fulfilled                no      EUR   
46  #1224             paid          fulfilled                no      EUR   
47  #1223             paid          fulfilled                no      EUR   
69  #1214             paid          fulfilled                no      EUR   
71  #1212             paid          fulfilled                no      EUR   
78  #1207             paid          fulfilled                no      EUR   

    Subtotal  Shipping  Taxes   Total  Discount Amount Shipping Method  \
5      38.94       0.0    0.0   38.94              0.0           Other   
24    200.10       0.0    0.0  200.10              0.0           Other   
36     54.50       0.0    0.0   54.50              0.0           Other   
46     26.16       0.0    0.0   26.16              0.0           Other   
47     40.94       0.0    0.0   40.94              0.0           Other   
69     51.15       0.0    0.0   51.15              0.0           Other   
71     32.65       0.0    0.0   32.65              0.0           Other   
78     27.50       0.0    0.0   27.50              0.0           Other   

   Billing City Billing Zip Payment Method  Refunded Amount            Id  \
5           NaN         NaN         manual              0.0  8.865550e+11   
24          NaN         NaN         manual              0.0  8.788850e+11   
36          NaN         NaN         manual              0.0  8.776710e+11   
46          NaN         NaN         manual              0.0  8.739650e+11   
47          NaN         NaN         manual              0.0  8.700700e+11   
69          NaN         NaN         manual              0.0  8.617570e+11   
71          NaN         NaN         manual              0.0  8.548820e+11   
78          NaN         NaN         manual              0.0  8.522650e+11   

   Risk Level               Source  Discount Code  \
5         Low  shopify_draft_order              0   
24        Low  shopify_draft_order              0   
36        Low  shopify_draft_order              0   
46        Low  shopify_draft_order              0   
47        Low  shopify_draft_order              0   
69        Low  shopify_draft_order              0   
71        Low               iphone              0   
78        Low  shopify_draft_order              0   

                                    Notes  
5                                    None  
24                  A entregar 6 de abril  
36                                   None  
46                                   None  
47                                   None  
69                                   None  
71  Es un pedido para Vimet (evento Wado)  
78    Es un pedido para Vimet (Coworking)

In almost all cases, when the order is delivered by an employee, the Billing City and Zip are null. We will replace this with 'Unknown'.

In [36]:
order_specific_clean['Billing City'] = order_specific_clean['Billing City'].fillna(value='Unknown')

In [37]:
order_specific_clean['Billing Zip'] = order_specific_clean['Billing Zip'].fillna(value='Unknown')

In [38]:
#We check the status of the NaN values
order_specific_clean.isna().sum()

Name                  0
Financial Status      0
Fulfillment Status    0
Accepts Marketing     0
Currency              0
Subtotal              0
Shipping              0
Taxes                 0
Total                 0
Discount Amount       0
Shipping Method       0
Billing City          0
Billing Zip           0
Payment Method        0
Refunded Amount       0
Id                    0
Risk Level            0
Source                0
Discount Code         0
Notes                 0
dtype: int64

We solved all the Nan values issues.

### Order-specific data: Data cleaning

Now we don't have NaN values but we need to check the content of the variables in order to clean it.

In [39]:
order_specific_clean.head()

Name Financial Status Fulfillment Status Accepts Marketing Currency  \
0  #1244             paid        unfulfilled               yes      EUR   
1  #1243             paid          fulfilled               yes      EUR   
3  #1242         refunded        unfulfilled               yes      EUR   
5  #1241             paid          fulfilled                no      EUR   
6  #1240             paid          fulfilled               yes      EUR   

   Subtotal  Shipping  Taxes  Total  Discount Amount Shipping Method  \
0     45.00       4.9    0.0  49.90              0.0  Envío estándar   
1     40.10       4.9    0.0  45.00              0.0  Envío estándar   
3     40.10       4.9    0.0  45.00              0.0  Envío estándar   
5     38.94       0.0    0.0  38.94              0.0           Other   
6     65.35       4.9    0.0  70.25             10.0  Envío estándar   

  Billing City Billing Zip Payment Method  Refunded Amount            Id  \
0    Barcelona      '08032         Stripe              0.0  8.910100e+11   
1    Barcelona      '08021         Stripe              0.0  8.869630e+11   
3    Barcelona      '08021         Stripe             45.0  8.869480e+11   
5      Unknown     Unknown         manual              0.0  8.865550e+11   
6   Barcelona       '08037         Stripe              0.0  8.853700e+11   

  Risk Level               Source  Discount Code  \
0        Low                  web              0   
1        Low                  web              0   
3        Low                  web              0   
5        Low  shopify_draft_order              0   
6        Low                  web              1   

                                               Notes  
0                                               None  
1                                               None  
3                                               None  
5                                               None  
6  ¿Podemos cambiar la lechuga larga por cualquie...

#### Name -- OK

We have checked **'Name'** previously and looks OK for now, additionally, we might drop ot to preserve 'ID', we¡ll decide this later

#### Financial Status -- OK

In [40]:
order_specific_clean['Financial Status'].value_counts()

paid                  146
partially_refunded     37
voided                 23
pending                 8
refunded                5
Name: Financial Status, dtype: int64

Looks good

#### Fulfillment Status -- OK

In [41]:
order_specific_clean['Fulfillment Status'].value_counts()

fulfilled      196
unfulfilled     22
partial          1
Name: Fulfillment Status, dtype: int64

Also good

#### Accepts Marketing

In [42]:
order_specific_clean['Accepts Marketing'].value_counts()

yes    184
no      35
Name: Accepts Marketing, dtype: int64

We will convert this to boolean (although, keeping int as type) to minimise work during the analysis later.

In [43]:
order_specific_clean['Accepts Marketing'] = order_specific_clean['Accepts Marketing'].map(lambda x: 1 if x=='yes' else 0)

#### Currency

In [44]:
order_specific_clean['Currency'].value_counts()

EUR    219
Name: Currency, dtype: int64

We drop currency as all is EUR and it is irrelevant for the analysis.

In [45]:
order_specific_clean = order_specific_clean.drop(columns='Currency')

#### Subtotal -- OK

In [46]:
order_specific_clean['Subtotal'].dtype

dtype('float64')

Also looks good

#### Shipping -- OK

In [47]:
order_specific_clean['Shipping'].value_counts()

4.9    129
0.0     86
3.9      4
Name: Shipping, dtype: int64

Also looks good

#### Taxes -- OK

In [48]:
order_specific_clean['Taxes'].value_counts()

0.00     213
10.40      1
2.72       1
53.80      1
17.34      1
1.21       1
8.04       1
Name: Taxes, dtype: int64

Some values are oddly high. Let's check the non-zero values

In [49]:
order_specific_clean[order_specific_clean['Taxes'] != 0.0]

Name    Financial Status Fulfillment Status  Accepts Marketing  \
96    #1199                paid          fulfilled                  1   
134   #1176              voided        unfulfilled                  1   
142   #1174              voided        unfulfilled                  1   
1746  #1075                paid          fulfilled                  1   
1752  #1074  partially_refunded          fulfilled                  0   
1870  #1065                paid          fulfilled                  1   

      Subtotal  Shipping  Taxes   Total  Discount Amount Shipping Method  \
96      310.00       0.0  53.80  310.00              0.0  Envío gratuito   
134     103.65       0.0  17.34  103.65              0.0  Envío gratuito   
142      66.75       4.9  10.40   71.65              0.0  Envío estándar   
1746     71.20       4.9   8.04   71.20              4.9  Envío estándar   
1752     40.35       4.9   2.72   40.35              4.9  Envío estándar   
1870     44.10       4.9   1.21   49.00              0.0  Envío estándar   

     Billing City Billing Zip Payment Method  Refunded Amount            Id  \
96      Barcelona      '08006         Stripe             0.00  8.374110e+11   
134     Barcelona      '08021         custom             0.00  8.154090e+11   
142     Barcelona      '08006         custom             0.00  8.116200e+11   
1746   Bellaterra      '08193         Stripe             0.00  5.015500e+11   
1752    Barcelona      '08021         Stripe             2.85  5.012710e+11   
1870    Barcelona      '08021         Stripe             0.00  4.844380e+11   

     Risk Level Source  Discount Code Notes  
96          Low    web              0  None  
134         Low    web              0  None  
142         Low    web              0  None  
1746        Low    web              1  None  
1752        Low    web              1  None  
1870        Low    web              0  None

Looks OK

#### Total <-- NEED TO CHECK WITH RAMON

In [61]:
order_specific_clean[order_specific_clean['Total'] != order_specific_clean['Subtotal'] + order_specific_clean['Shipping']]

Name    Financial Status Fulfillment Status  Accepts Marketing  \
25    #1230                paid          fulfilled                  1   
48    #1222                paid          fulfilled                  1   
95    #1201                paid          fulfilled                  1   
103   #1197                paid          fulfilled                  1   
114   #1191                paid          fulfilled                  1   
116   #1189                paid          fulfilled                  0   
117   #1188                paid          fulfilled                  1   
127   #1180                paid          fulfilled                  1   
133   #1177                paid          fulfilled                  1   
145   #1172              voided        unfulfilled                  1   
601   #1149                paid          fulfilled                  1   
665   #1145              voided        unfulfilled                  1   
921   #1126                paid          fulfilled                  0   
951   #1123              voided          fulfilled                  1   
1051  #1117                paid          fulfilled                  1   
1110  #1114                paid          fulfilled                  1   
1124  #1113  partially_refunded          fulfilled                  1   
1161  #1111                paid          fulfilled                  0   
1257  #1106                paid          fulfilled                  1   
1276  #1105                paid          fulfilled                  1   
1284  #1104  partially_refunded          fulfilled                  1   
1323  #1102  partially_refunded          fulfilled                  1   
1449  #1095                paid          fulfilled                  1   
1525  #1091                paid          fulfilled                  1   
1542  #1090                paid          fulfilled                  1   
1559  #1089                paid          fulfilled                  1   
1576  #1087                paid          fulfilled                  1   
1592  #1086                paid          fulfilled                  1   
1629  #1084                paid          fulfilled                  1   
1709  #1078                paid          fulfilled                  1   
1746  #1075                paid          fulfilled                  1   
1752  #1074  partially_refunded          fulfilled                  0   
1764  #1073                paid          fulfilled                  1   
1800  #1071                paid          fulfilled                  1   
1817  #1069                paid          fulfilled                  1   
1827  #1068                paid          fulfilled                  1   
1876  #1064                paid          fulfilled                  1   
2031  #1053  partially_refunded          fulfilled                  1   
2138  #1040                paid          fulfilled                  1   
2214  #1032  partially_refunded            partial                  1   

      Subtotal  Shipping  Taxes  Total  Discount Amount Shipping Method  \
25       66.05       4.9   0.00  66.05              4.9          shargo   
48       38.05       4.9   0.00  42.95              0.0          shargo   
95       59.00       4.9   0.00  59.00              4.9          shargo   
103      32.50       4.9   0.00  32.50              4.9          shargo   
114      32.50       4.9   0.00  32.50              4.9          shargo   
116      65.00       4.9   0.00  65.00              4.9          shargo   
117      65.00       4.9   0.00  65.00              4.9          shargo   
127      59.95       4.9   0.00  64.85              0.0          shargo   
133      59.95       4.9   0.00  64.85              0.0          shargo   
145       1.40       4.9   0.00   6.30              0.0          shargo   
601      61.55       4.9   0.00  61.55              4.9          shargo   
665      65.90       4.9   0.00  70.80              0.0          shargo   
921       9.00       4.9   0.00   9.00

When there is a Discount regarding free shipping, the discount is applied directly to the shipment cost.

WARNING! Sometimes, the above query is not fulfilled but pandas returns the rows ??

#### Discount Amount <-- OK

In [51]:
order_specific_clean['Discount Amount'].value_counts()

0.0     183
4.9      24
10.0     11
32.4      1
Name: Discount Amount, dtype: int64

In [52]:
order_specific_clean['Taxes'].dtype

dtype('float64')

Looks good

#### Shipping Method

In [59]:
order_specific_clean['Shipping Method'].value_counts()

shargo      137
gratuito     51
mercado      21
otros        10
Name: Shipping Method, dtype: int64

Looks good but we will replace the values to reduce the amount of text. This can be useful later if we want to display this in a plot

In [54]:
order_specific_clean['Shipping Method'] = order_specific_clean['Shipping Method'].map(lambda x: 'shargo' if x=='Envío estándar' else ('gratuito' if x == 'Envío gratuito' else ('mercado' if x == 'Recogida en el mercado' else ('shargo' if x == 'El reparto se realiza a través de Shargo' else 'otros'))))

#### 	Billing City

In [56]:
order_specific_clean['Billing City'].value_counts()

Barcelona                  166
Barcelona                   16
barcelona                   14
Unknown                      8
test                         7
BARCELONA                    2
jsjajsjsjjs                  2
khg                          1
Bellaterra                   1
Sant feliu de Llobregat      1
dsa                          1
Name: Billing City, dtype: int64

In [57]:
order_specific_clean = order_specific_clean.replace(['Barcelona ', ' Barcelona', 'barcelona', 'BARCELONA'], 'Barcelona')
order_specific_clean = order_specific_clean.replace(['khg', 'dsa', 'test', 'jsjajsjsjjs'], 'Unknown')

In [58]:
order_specific_clean['Billing City'].value_counts()

Barcelona                  198
Unknown                     19
Bellaterra                   1
Sant feliu de Llobregat      1
Name: Billing City, dtype: int64

#### Billing Zip

#### Payment Method

#### Refunded Amount

#### Id

#### Risk Level

#### Source

#### Discount Code

#### Notes